### PR TITLE
Assignement4 - Spencer West & Matthew Powers

### PR DESCRIPTION
**Pull request by SaaWest**
@SaaWest @Matthew-Powers

### CODE REVIEW

**Comment by colestewart12** [comment]
I can review this!

File: Logger.Tests/FullNameTests.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+            FullName fullName = new(firstName, lastName, middleName);
+
+            Assert.Equal("Michael Jordon", fullName.Name);
+            //Assert.Equal(fullName.MiddleName, middleName);
+            //Assert.Equal(fullName.FirstName, lastName);
+
+        }
+        [Fact]
+        public void FullName_FirstName_IsNull()
```
Comment (colestewart12) [inline Logger.Tests/FullNameTests.cs:45]: The test name is a little confusing...
```suggestion
        public void FullName_MissingLastName_IsNull()
```

File: Logger/FullName.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+﻿namespace Logger;
+
+
+public record FullName
+{
+    //Fullname encompasses three parts usually a first, middle and last name a constructor with 
+    //these three/two requirements is what is need when instantiate a new person
+    //Records are immutable and so FullName should reflect that as first, middle, and last names do
+    //not change in real life unless some legal entity supercedes that.
```
Comment (colestewart12) [inline Logger/FullName.cs:9]: Make sure to comment whether FullName is value or reference type

File: Logger.Tests/EmployeeRecordTests.cs
Comment (colestewart12) [inline Logger.Tests/EmployeeRecordTests.cs:1]: More test coverage is needed for Employee

**Review by colestewart12 — COMMENTED** [review]
Define a full name record (first, last, middle) handling optional and null appropriately. ✔
Provide a comment on the full name record on why you selected to define a value or a reference type and ❌
Provide a comment on the full name record on why or why not the type is immutable. ✔
Define an IEntity interface: ✔
Add an Id property of type Guid that is an init-only setter. ✔
Add a Name property that is string. ✔
Define an abstract base class that implements IEntity - appropriately choosing to implement the interface explicitly or implicitly. ✔
Do not implement the Name property in this abstract class. ✔
Do force any derived classes to provide an implementation for Name. ✔
Define book, student, and employee records - all with entity capabilities. ❌(Still working on)
Choose a thoughtful approach for the value of Name for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ❌(Still working on)
Provide a comment on each interface member in each entity explaining why you implemented it implicitly or explicitly. ❌
You should consider the relationship between Student and Employee and refactor the common code shared between them. ❌
Given the properties on the full name record, you should consider which entities it makes sense to use it. ❌
Test that the equality behavior on these entities behaves as expected. ❌
Provide a comment on each interface method explaining why you implemented it implicitly or explicitly.❌
Unit test the Storage class using various entities to ensure that there are no bugs.❌
Pay special attention to the CreateLogger methods throughout.
Fundamentals
Be sure you enable:
Enable Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Ensure that you turn on Nullability (Nullable) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ❌
Choose simplicity over complexity ✔

File: Logger.Tests/BookRecordTests.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    public void BookRecord_With_BookName_AuthorName_NotNull()
+    {
+        BookRecord book = new("Do Robots Dream of Electric Sheep", "Philip K. Dick");
+        Assert.NotNull(book);
+        Assert.Equal("Do Robots Dream of Electric Sheep", book.BookName);
+        Assert.Equal("Philip K. Dick", book.AuthorName);
+    }
+
+    [Fact]
+    public void BookRecord_With_BookName_Null()
```
[the remark this answers is not in the document]
Comment (SaaWest) [reply Logger.Tests/BookRecordTests.cs:16]: will change it

File: Logger.Tests/BookRecordTests.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        Assert.Throws<ArgumentNullException>(() => new BookRecord(null!, "Philip k. Dick"));
+
+
+    }
+    [Fact]
+    public void BookRecord_With_AuthorName_Null()
```
[the remark this answers is not in the document]
Comment (SaaWest) [reply Logger.Tests/BookRecordTests.cs:23]: will can it

File: Logger.Tests/EmployeeRecordTests.cs
Comment (SaaWest) [reply Logger.Tests/EmployeeRecordTests.cs:1]: I don't understand how that happened but these are old tests and were refactored to have four testing parameters
ex
`    public void EmployeeRecord_MiddleName_Null()
    {
        EmployeeRecord employee = new EmployeeRecord("Mike", "Soft", null!);
        Assert.Null(employee.MiddleName);
        Assert.Equal("Mike Soft", employee.writeName);

    }
    [Fact]
    public void FullName_FirstName_IsNull()
    {
        string firstName = "Steven";
        Assert.Throws<ArgumentNullException>(() => new EmployeeRecord(firstName, null!));



    }`

File: Logger.Tests/FullNameTests.cs (hunk 1 above)
Comment (SaaWest) [reply Logger.Tests/FullNameTests.cs:45]: Will change... thank you

File: Logger.Tests/Logger.Tests.csproj [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+    <IsPackable>false</IsPackable>
+    <Nullable>enable</Nullable>
+    <LangVersion>12.0</LangVersion>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <EnableNETAnalyzers>true</EnableNETAnalyzers>
+    <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```
[the remark this answers is not in the document]
Comment (SaaWest) [reply Logger.Tests/Logger.Tests.csproj:11]: Thanks

File: Logger/BookRecord.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    public BookRecord(string bookName, string authorName)
+    {
+        BookName = bookName ?? throw new ArgumentNullException(nameof(bookName));
+        AuthorName = authorName ?? throw new ArgumentNullException(nameof(authorName));
+
+    }
+    public string BookName { get; set; }
+    public string AuthorName { get; set; }
+    public Guid ID { init => throw new NotImplementedException(); }
+    public string Name { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }
```
[the remark this answers is not in the document]
Comment (SaaWest) [reply Logger/BookRecord.cs:14]: Working on it.. trying to figure it out

File: Logger/EmployeeRecord.cs [hunk 7]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Logger;
+
+public record EmployeeRecord : IEntity
+{
+    public Guid ID { init => throw new NotImplementedException(); }
+    public string Name { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }
```
[the remark this answers is not in the document]
Comment (SaaWest) [reply Logger/EmployeeRecord.cs:6]: This portion is also out of date
`    
public EmployeeRecord(string firstName, string lastName, string? middleName = null )
    {
        FirstName = string.IsNullOrEmpty(firstName) ? throw new ArgumentNullException(nameof(firstName)) : firstName;
        LastName = string.IsNullOrEmpty(lastName) ? throw new ArgumentNullException(nameof(lastName)) : lastName;
        MiddleName = middleName;

    }
    public string FirstName { get; set; }
    public string LastName { get; set; }
    public string? MiddleName { get; set; }
    public string writeName { get { return FirstName + $"{(MiddleName != null ? " " + MiddleName + " " : " ")}" + LastName; } }
    public Guid ID { init => throw new NotImplementedException(); }
    public string Name { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }`

File: Logger/FullName.cs (hunk 2 above)
Comment (SaaWest) [reply Logger/FullName.cs:9]: done

File: Logger/BookRecord.cs [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+public record BookRecord : IEntity
+{
+    public BookRecord(string bookName, string authorName)
+    {
+        BookName = bookName ?? throw new ArgumentNullException(nameof(bookName));
+        AuthorName = authorName ?? throw new ArgumentNullException(nameof(authorName));
+
+    }
+    public string BookName { get; set; }
+    public string AuthorName { get; set; }
```
[the remark this answers is not in the document]
Comment (SaaWest) [reply Logger/BookRecord.cs:14]: Thank you

**Pull request closed without merging**